services/api.py

```diff
@@ -816,20 +816,38 @@ def service_nodes_by_ancestors(service_node_ids):
                 bbox_geometry_filter = munigeo_api.build_bbox_filter(ref, val, 'geometry')
                 queryset = queryset.filter(Q(**bbox_filter) | Q(**bbox_geometry_filter))
 
+        def get_category_ids(categories, category):
+            ids = []
+            for mapping in categories:
+                if mapping.split(':')[0] == category:
+                    ids.append(mapping.split(':')[1])
+            return ids
+
         if 'category' in filters:
             services_and_service_nodes = filters.get('category', None).split(',')
-            service_ids = []
-            servicenode_ids = []
-            for category in services_and_service_nodes:
-                key = category.split(':')[0]
-                value = category.split(':')[1]
-                if key == 'service':
-                    service_ids.append(value)
-                elif key == 'service_node':
-                    servicenode_ids.append(value)
+            service_ids = get_category_ids(services_and_service_nodes, 'service')
+            servicenode_ids = get_category_ids(services_and_service_nodes, 'service_node')
             queryset = queryset.filter(Q(services__in=service_ids)
                                        | Q(service_nodes__in=service_nodes_by_ancestors(servicenode_ids))).distinct()
 
+        if 'period' in filters:
+            period = filters.get('period', None)
+            services = None
+            if 'service' in filters:
+                services = filters.get('service').split(',')
+            if 'category' in filters:
+                categories = filters.get('category', None).split(',')
+                services = get_category_ids(categories, 'service')
```

Comment:
I forgot to mention that also service_nodes can have periods. They do when any of the related services do.

Reply:
Hmm.. this might be hard to implement. Let's spend some time designing this.

Reply:
There are two options.
1 reimplement all service_node filters by translating them to service filters
2 extract period-enabled services from the service_nodes and use them for the period filter, and the service_nodes themselves for the service_node filter.

Reply:
Since the service_node filter currently takes the whole tree hierarchy into account, option 1 might be difficult.

Reply:
This is simpler and better than what I had in mind. It seems to work with the assumption that

    if a ServiceNode has periods enabled, all the subexpressions in its service_reference expression
    also have periods enabled

This seems to be the case. Maybe you could add the above assumption as a comment. If the assumption doesn't hold, this filter could be too restrictive, because the unit could match a subexpression which doesn't have a period restriction, but this would still filter by the other subexpressions which do.